controllers/lease_controller.go

```diff
@@ -0,0 +1,544 @@
+		if etcdBackupImage == "" {
+			return nil, fmt.Errorf("either etcd resource or image vector should have %s image", common.BackupRestore)
+		}
+	} else {
+		etcdBackupImage = *etcd.Spec.Backup.Image
+	}
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getJobName(etcd),
+			Namespace: etcd.Namespace,
+			Labels:    getLabels(etcd),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         "druid.gardener.cloud/v1alpha1",
+					BlockOwnerDeletion: pointer.BoolPtr(true),
+					Controller:         pointer.BoolPtr(true),
+					Kind:               "Etcd",
+					Name:               etcd.Name,
+					UID:                etcd.UID,
+				},
+			},
+		},
+
+		Spec: batchv1.JobSpec{
+			ActiveDeadlineSeconds: pointer.Int64Ptr(int64(activeDeadlineSeconds)),
+			Completions:           pointer.Int32Ptr(1),
+			BackoffLimit:          pointer.Int32Ptr(0),
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: etcd.Spec.Annotations,
+					Labels:      getLabels(etcd),
+				},
+				Spec: v1.PodSpec{
+					ActiveDeadlineSeconds: pointer.Int64Ptr(int64(activeDeadlineSeconds)),
+					ServiceAccountName:    getServiceAccountName(etcd),
+					RestartPolicy:         v1.RestartPolicyNever,
+					Containers: []v1.Container{{
+						Name:            "compact-backup",
+						Image:           etcdBackupImage,
+						ImagePullPolicy: v1.PullIfNotPresent,
+						Command:         getCompactJobCommands(etcd),
+						VolumeMounts:    getCmpctJobVolumeMounts(etcd, logger),
+						Env:             getCmpctJobEnvVar(etcd, logger),
+					}},
+					Volumes: getCmpctJobVolumes(etcd, logger),
+				},
+			},
+		},
+	}
+
+	if etcd.Spec.Backup.CompactionResources != nil {
+		job.Spec.Template.Spec.Containers[0].Resources = *etcd.Spec.Backup.CompactionResources
+	}
+
+	logger.Info("Creating job", "job", kutil.Key(job.Namespace, job.Name).String())
+	err = lc.Create(ctx, job)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO (abdasgupta): Evaluate necessity of claiming object here after creation
+	return job, nil
+}
+
+func getCronJobName(etcd *druidv1alpha1.Etcd) string {
+	return fmt.Sprintf("%s-compact-backup", etcd.Name)
+}
+
+func getJobName(etcd *druidv1alpha1.Etcd) string {
+	return fmt.Sprintf("%s-compact-job", string(etcd.UID[:6]))
+}
+
+func getLabels(etcd *druidv1alpha1.Etcd) map[string]string {
+	return map[string]string{
+		"name":                             "etcd",
+		"instance":                         etcd.Name,
+		"garden.sapcloud.io/role":          "controlplane",
+		"networking.gardener.cloud/to-dns": "allowed",
+		"networking.gardener.cloud/to-private-networks": "allowed",
+		"networking.gardener.cloud/to-public-networks":  "allowed",
+	}
+}
+func getCmpctJobVolumeMounts(etcd *druidv1alpha1.Etcd, logger logr.Logger) []v1.VolumeMount {
+	vms := []v1.VolumeMount{
+		{
+			Name:      "etcd-workspace-dir",
+			MountPath: "/var/etcd/data",
+		},
+	}
+
+	if etcd.Spec.Backup.Store == nil {
```

Comment:
When would this condition be true? Compaction job is reconciled only when `etcd.Spec.Backup.Store != nil` right? And we know that this field is technically immutable now, so it can't change mid-reconciliation.

Reply:
It is to avoid any unintended nil pointer error due to any unknown circumstances(a particular memory address might be wiped out for unknown reason). It would even help to write future unit test cases where developer unknowingly changed the value to nil at some place else.